scripts/tutorial-runner.ts

```diff
@@ -0,0 +1,174 @@
+import { hideBin } from 'yargs/helpers';
+import sh from 'shelljs';
+import fs from 'fs';
+import path from 'path';
+
+type CodeBlock = ShellCommands | CodePatch;
+
+type ShellCommands = {
+  lang: 'sh';
+  commands: string[];
+};
+
+type CodePatch = {
+  lang: 'ts';
+  filePath: string;
+  startLineNum: number;
+  codeLines: string[];
+};
+
+yargs(hideBin(process.argv))
+  .command(
+    '$0',
+    'Execute all of the code blocks in a tutorial in sequence, ensuring that each shell command executed returns a non-zero exit code.',
+    {},
+    (argv) => {
+      if (process.env.GITHUB_ACTIONS !== 'true')
+        throw 'This is only intended to be run in a GitHub Actions environment.';
+
+      if (argv._[0] === undefined || typeof argv._[0] !== 'string') {
+        throw "Provide the tutorial's file path as an argument";
+      }
+      const tutorialPath = argv._[0];
+      const markdown = sh.cat(tutorialPath);
+
+      const testDir = `test-${tutorialPath.replace(/\/|\./g, '-')}`;
+      const pwd = sh.pwd().toString();
+
+      try {
+        sh.mkdir(testDir);
+        sh.cd(testDir);
+
+        // 1. Extract code blocks.
+        const regex =
+          /(?<=```)(?<infoString>[\w\/\. ]+?)\n(?<code>.+?)\n(?=```)/gs;
+        const codeBlocks = Array.from(markdown.matchAll(regex)).map(
+          regexMatchToCodeBlock
+        );
+
+        // 2. Simulate tutorial
+        codeBlocks.forEach((codeBlock) => {
+          if (codeBlock.lang === 'sh') {
+            executeShellCommand(codeBlock);
+          } else if (codeBlock.lang === 'ts') {
+            applyCodePatch(codeBlock);
+          }
+        });
+      } finally {
+        sh.rm('-r', path.join(pwd, testDir));
+      }
+    }
+  )
+  .parse();
+
+function regexMatchToCodeBlock(match: RegExpMatchArray): CodeBlock {
+  // The "info string" (https://spec.commonmark.org/0.30/#info-string) stores metadata about the code block.
+  const infoString = match.groups.infoString;
+  const infoStringSegments = infoString.split(' ');
+  const [lang] = infoStringSegments;
+  const code = match.groups.code;
+
+  if (lang === 'ts') {
+    const filePath: string | undefined = infoStringSegments[1];
+
+    if (filePath) {
+      const getLineNum = (codeLineWithNum: string): number =>
+        parseInt(codeLineWithNum.match(/\d+/)[0]);
+      const stripLineNum = (codeLineWithNum: string): string => {
+        if (codeLineWithNum.match(/\d+$/)) {
+          // If there's only a line number, the rest of the line is empty
+          return '';
+        } else {
+          return codeLineWithNum.substring(codeLineWithNum.indexOf(' ') + 1);
+        }
+      };
+      const codeLinesWithNums = code.split('\n');
+      const startLineNum = getLineNum(codeLinesWithNums[0]);
+      const codeLines = codeLinesWithNums.map((codeLineWithNum, index) => {
+        if (startLineNum + index !== getLineNum(codeLineWithNum)) {
+          throw (
+            'Line numbers must be contiguous per code block.\n' +
```

Comment:
It's fine for now, but I think it could be nice to remove this restriction in the future. It offers the possibility to leave out stuff in the code samples, to be more concise but still provide valuable context, like
```
4   class MyContract extends SmartContract {
...
9     @method myMethod() {
10      // some code we care about
11    }
12  }
```

Reply:
Good point, and that snippet in particular is a great example. I was originally thinking that each range could just be a different code block, but I think that would be weird and possibly confusing in the snippet you provided. I can look into updating this as I set this up for other tutorials!